src/platform/utilities/ui/webComponents.js

```diff
@@ -0,0 +1,166 @@
+
+    if (isWebComponentReady(host) || process.env.NODE_ENV === 'test') {
+      // shadowRoot not populated in React testing library so just return
+      resolve(host);
+      return;
+    }
+
+    if (host.hasAttribute('data-observing-shadow')) {
+      return;
+    }
+    host.setAttribute('data-observing-shadow', 'true');
+
+    const hostObserver = new MutationObserver(() => {
+      try {
+        if (isWebComponentReady(host)) {
+          // shadowRoot will exist, but its contents may not be
+          // visible at this point because it hasn't rerendered
+          if (waitForPaint) {
+            requestAnimationFrame(() => {
+              resolve(host);
+            });
+          } else {
+            resolve(host);
+          }
+          host.removeAttribute('data-observing-shadow');
+          if (hostObserver) {
+            hostObserver.disconnect();
+          }
+        }
+      } catch (error) {
+        reject(new Error(`An error occurred in waitForShadowRoot`));
+        console.error('An error occurred in waitForShadowRoot: ', error);
+        host.removeAttribute('data-observing-shadow');
+        if (hostObserver) {
+          hostObserver.disconnect();
+        }
+      }
+    });
+
+    hostObserver.observe(host, {
+      childList: true,
+      attributes: true,
+      subtree: true,
+    });
+  });
+}
+
+/**
+ * An async querySelector that waits for all necessary shadowRoots involved
+ *
+ * ```
+ * // Example with await:
+ * const el = await querySelectorWithShadowRoot('va-checkbox#id-1')
+ * const el = await querySelectorWithShadowRoot('va-checkbox#id-1', 'va-checkbox-group')
+ * const el = await querySelectorWithShadowRoot('va-checkbox#id-1', container)
+ * console.log(el.shadowRoot)
+ *
+ * // Example with then:
+ * querySelectorWithShadowRoot('va-checkbox#id-1')
+ *    .then(el => console.log(el.shadowRoot))
+ * querySelectorWithShadowRoot('va-checkbox#id-1', 'va-checkbox-group')
+ *    .then(el => console.log(el.shadowRoot))
+ * querySelectorWithShadowRoot('va-checkbox#id-1', container)
+ *    .then(el => console.log(el.shadowRoot))
+ * ```
+ *
+ * @param {string | HTMLElement} selector
+ * @param {HTMLElement} [root]
+ * @returns {Promise<HTMLElement | null>}
+ */
+export async function querySelectorWithShadowRoot(selector, root) {
+  try {
+    let element = typeof selector === 'string' ? null : selector;
+    let currentRoot =
+      typeof root === 'string'
+        ? document.querySelector(root)
+        : root || document;
+
+    if (isWebComponent(currentRoot) && !isWebComponentReady(currentRoot)) {
+      const waitForPaint = false;
+      await waitForShadowRoot(currentRoot, waitForPaint);
+      // Fallback to root if shadowRoot is not available,
+      // for example in unit tests where shadowRoot is null
+      currentRoot = currentRoot.shadowRoot || currentRoot;
+    }
+
+    if (!element) {
+      element = currentRoot.querySelector(selector);
+    }
```

Comment:
I feel like this could be rewritten here as:

```js
const element = typeof selector === 'string' ? currentRoot.querySelector(selector) : selector;
```

instead of at the beginning

Reply:
That does seem better. Updated.